renderdoc/driver/ihv/amd/amd_counters.cpp

```diff
@@ -22,70 +22,202 @@
+    dllName += "DX11";
+    break;
+  case AMDCounters::eApiType_Dx12:
+    dllName += "DX12";
+    break;
+  case AMDCounters::eApiType_Ogl:
+    dllName += "GL";
+    break;
+  case AMDCounters::eApiType_Vk:
+    dllName += "VK";
+    break;
+  default:
+    RDCWARN(
+      "AMD GPU performance counters could not be initialized successfully. "
+      "Unsupported API type specified");
+    return false;
+  }
+
+#if ENABLED(RDOC_WIN32)
+#if ENABLED(RDOC_X64)
+  dllName += "-x64";
+#endif
+  dllName += ".dll";
+#else
+  dllName = "lib" + dllName;
+  dllName += ".so";
+#endif
+
 
+
+  GPA_GetFuncTablePtrType getFuncTable = NULL;
+
+#if ENABLED(RDOC_WIN32)
   // first try in the plugin location it will be in distributed builds
-  HMODULE module = LoadLibraryA(LocatePluginFile("amd/counters", dllName).c_str());
+  HMODULE module = LoadLibraryA(LocatePluginFile("amd/counters", dllName.c_str()).c_str());
 
   // if that failed then try checking for it just in the default search path
   if(module == NULL)
   {
-    module = LoadLibraryA(dllName);
+    module = LoadLibraryA(dllName.c_str());
   }
 
   if(module == NULL)
   {
     RDCWARN(
         "AMD GPU performance counters could not be initialized successfully. "
         "Are you missing the DLLs?");
-
     return false;
   }
 
-  m_pGPUPerfAPI = new GPAApi();
-  GPA_GetFuncTablePtrType getFuncTable =
-      (GPA_GetFuncTablePtrType)GetProcAddress(module, "GPA_GetFuncTable");
+  getFuncTable = (GPA_GetFuncTablePtrType)GetProcAddress(module, "GPA_GetFuncTable");
+#else
+  // first try in the plugin location it will be in distributed builds
+  void* module = dlopen(LocatePluginFile("amd/counters", dllName.c_str()).c_str(), RTLD_LAZY);
 
+    // if that failed then try checking for it just in the default search path
+    if(module == NULL)
+    {
+      module = dlopen(dllName.c_str(), RTLD_LAZY);
+    }
+
+    if(module == NULL)
+    {
+      RDCWARN(
+          "AMD GPU performance counters could not be initialized successfully. "
+          "Are you missing the DLLs?");
+      return false;
+    }
+
+    getFuncTable = (GPA_GetFuncTablePtrType)dlsym(module, "GPA_GetFuncTable");
+#endif
+
+  m_pGPUPerfAPI = new GPAApi();
   if(getFuncTable)
   {
     getFuncTable((void **)&m_pGPUPerfAPI);
   }
   else
   {
-    RDCWARN(
-        "GPA version is out of date, doesn't expose GPA_GetFuncTable. Make sure you have 3.0 or "
-        "above.");
     delete m_pGPUPerfAPI;
-    m_pGPUPerfAPI = NULL;
+    GPA_LoggingCallback(GPA_LOGGING_ERROR, "Failed to get GPA function table. Invalid dynamic library?");
+    return false;
+  }
+
+  GPA_Logging_Type loggingType = GPA_LOGGING_ERROR;
+#if ENABLED(RDOC_DEVEL)
+  loggingType = GPA_LOGGING_ERROR_AND_MESSAGE;
+#endif
+  auto status = m_pGPUPerfAPI->GPA_RegisterLoggingCallback(loggingType, GPA_LoggingCallback);
+  if (AMD_FAILED(status))
+  {
+    GPA_LoggingCallback(GPA_LOGGING_ERROR, FormatErrMessage("Failed to initialize logging", status).c_str());
```

Comment:
It seems a little convoluted to go via `GPA_LoggingCallback` and as discussed there it should probably just forward this to `RDCERR`, so I think instead it'd be better to convert to something like:

```cpp
RDCERR("Failed to initialize logging. %s", m_pGPUPerfAPI->GPA_GetStatusAsStr(status));
```

If you wanted you could define a local macro to remove repetition a bit.